components/txn_types/src/lock.rs

```diff
@@ -461,126 +641,180 @@ impl Lock {
+                        let lock_bytes = bytes::decode_compact_bytes(&mut b)?;
+                        let lock_ts = detect_lock_ts(&lock_bytes)?;
+                        segments.insert(lock_ts, Either::Left(lock_bytes));
+                    }
+                }
+                _ => {
+                    // To support forward compatibility, all fields should be serialized in order
+                    // and stop parsing if meets an unknown byte.
+                    break;
+                }
+            }
         }
 
-        // There is a pending lock. Client should wait or clean it.
-        Err(Error::from(ErrorInner::KeyIsLocked(
-            lock.into_owned().into_lock_info(raw_key),
-        )))
+        Ok(Self::new_with_txn_infos(segments))
     }
 
-    // Check if lock could be bypassed for isolation level `RcCheckTs`.
-    fn check_ts_conflict_rc_check_ts(
-        lock: Cow<'_, Self>,
-        key: &Key,
-        ts: TimeStamp,
-        bypass_locks: &TsSet,
-    ) -> Result<()> {
-        if lock.lock_type == LockType::Lock || lock.is_pessimistic_lock() {
-            // Ignore lock when the lock's type is Lock or Pessimistic.
-            return Ok(());
+    pub fn len(&self) -> usize {
+        self.txn_info_segments.len()
+    }
+
+    #[inline]
+    #[must_use]
+    pub fn contains_start_ts(&self, start_ts: TimeStamp) -> bool {
+        self.txn_info_segments.contains_key(&start_ts)
+    }
+
+    pub fn get_lock(&mut self, ts: &TimeStamp) -> Option<&Lock> {
+        if let Some(either) = self.txn_info_segments.get_mut(ts) {
+            match either {
+                Either::Left(encoded) => {
+                    let lock = Lock::parse(encoded).expect("failed to parse shared lock txn info");
+                    *either = Either::Right(lock);
+                    either.as_ref().right()
+                }
+                Either::Right(lock) => Some(lock),
+            }
+        } else {
+            None
         }
+    }
 
-        // The lock is resolved already.
-        if bypass_locks.contains(lock.ts) {
-            return Ok(());
+    // Decode all shared-lock segments so tests can rely on parsed locks.
+    #[cfg(test)]
+    pub fn parse_all(&mut self) {
+        for (_ts, either) in self.txn_info_segments.iter_mut() {
+            if let Either::Left(encoded) = either {
+                let lock = Lock::parse(encoded).expect("failed to parse shared lock txn info");
+                *either = Either::Right(lock);
+            }
         }
+    }
 
-        // Return conflict error.
-        Err(Error::from(ErrorInner::WriteConflict {
-            start_ts: ts,
-            conflict_start_ts: lock.ts,
-            conflict_commit_ts: Default::default(),
-            key: key.to_raw()?,
-            primary: lock.primary.to_vec(),
-            reason: WriteConflictReason::RcCheckTs,
-        }))
+    #[inline]
+    #[must_use]
+    pub fn find_shared_lock_txn(&mut self, start_ts: TimeStamp) -> Option<&Lock> {
```

Comment:
```suggestion
    pub fn get_shared_lock_txn(&mut self, start_ts: TimeStamp) -> Option<&Lock> {
```
is more consistent naming using `get_xxx` for query.